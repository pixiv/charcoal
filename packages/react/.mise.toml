[tools]
pnpm = "10.17.1"
