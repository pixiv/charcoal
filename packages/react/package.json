{
  "name": "@charcoal-ui/react",
  "version": "4.0.0-beta.3",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs.js",
      "import": "./dist/index.esm.js",
      "default": "./dist/index.esm.js"
    },
    "./dist/index.css": {
      "import": "./dist/index.css",
      "require": "./dist/index.css"
    }
  },
  "types": "./dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "run-p --print-label 'build:*'",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "typecheck": "tsc --project tsconfig.build.json --pretty --noEmit",
    "clean": "rimraf dist .tsbuildinfo"
  },
  "devDependencies": {
    "@charcoal-ui/esbuild-plugin-styled-components": "^4.0.0-beta.3",
    "@react-types/switch": "^3.1.2",
    "@storybook/addon-actions": "^7.4.1",
    "@storybook/addon-knobs": "^7.0.2",
    "@storybook/addons": "^7.4.1",
    "@storybook/api": "^7.4.1",
    "@storybook/components": "^7.4.1",
    "@storybook/core-events": "^7.4.1",
    "@storybook/react": "^7.4.1",
    "@storybook/react-webpack5": "^7.4.1",
    "@storybook/theming": "^7.4.1",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.0",
    "@types/jest-axe": "^3.5.3",
    "@types/react": "^17.0.38",
    "@types/react-dom": "^17.0.11",
    "@types/react-router-dom": "^5.3.3",
    "@types/styled-components": "^5.1.21",
    "@types/warning": "^3.0.0",
    "jest-axe": "^5.0.1",
    "jest-styled-components": "^7.0.8",
    "npm-run-all": "^4.1.5",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router-dom": "^6.2.1",
    "rimraf": "^3.0.2",
    "storybook": "^7.4.1",
    "styled-components": "^5.3.3",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@charcoal-ui/icons": "^4.0.0-beta.3",
    "@charcoal-ui/styled": "^4.0.0-beta.3",
    "@charcoal-ui/theme": "^4.0.0-beta.3",
    "@charcoal-ui/utils": "^4.0.0-beta.3",
    "@react-aria/button": "^3.9.1",
    "@react-aria/checkbox": "^3.13.0",
    "@react-aria/dialog": "^3.5.10",
    "@react-aria/focus": "^3.16.0",
    "@react-aria/link": "^3.6.5",
    "@react-aria/overlays": "^3.20.0",
    "@react-aria/radio": "^3.10.0",
    "@react-aria/ssr": "^3.9.1",
    "@react-aria/switch": "^3.6.0",
    "@react-aria/textfield": "^3.14.1",
    "@react-aria/utils": "^3.23.0",
    "@react-aria/visually-hidden": "^3.8.8",
    "polished": "^4.1.4",
    "react-spring": "^9.0.0",
    "react-stately": "^3.26.0",
    "warning": "^4.0.3"
  },
  "peerDependencies": {
    "react": ">=17.0.0",
    "styled-components": ">=5.1.1"
  },
  "files": [
    "src",
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/react"
  },
  "gitHead": "e1ece2e43901ae667afdd5c178040607d939dcd5"
}
