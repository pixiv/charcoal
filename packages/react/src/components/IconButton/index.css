.charcoal-icon-button {
  cursor: pointer;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;

  user-select: none;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 999999px;
  transition: 0.2s background-color, 0.2s box-shadow;
}

.charcoal-icon-button:disabled,
.charcoal-icon-button[data-variant]:disabled,
.charcoal-icon-button[aria-disabled]:not([aria-disabled='false']) {
  cursor: default;
  opacity: 0.32;
}
.charcoal-icon-button:focus {
  outline: none;
}
.charcoal-icon-button::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.charcoal-icon-button[data-size='XS'] {
  width: 20px;
  height: 20px;
}
.charcoal-icon-button[data-size='S'] {
  width: 32px;
  height: 32px;
}
.charcoal-icon-button[data-size='M'] {
  width: 40px;
  height: 40px;
}

.charcoal-icon-button[data-variant='Default'] {
  color: var(--charcoal-color-icon-tertiary-default);
  background-color: rgba(0, 0, 0, 0);
}
.charcoal-icon-button[data-variant='Default'][data-active='true']:not(
    :disabled
  ):not([aria-disabled]),
.charcoal-icon-button[data-variant='Default'][data-active='true'][aria-disabled='false'] {
  color: var(--charcoal-color-icon-tertiary-press);
  background-color: var(--charcoal-color-container-press-a);
}
.charcoal-icon-button[data-variant='Default'][data-active='false']:not(
    :disabled
  ):not([aria-disabled]):hover,
.charcoal-icon-button[data-variant='Default'][data-active='false'][aria-disabled='false']:hover {
  color: var(--charcoal-color-icon-tertiary-hover);
  background-color: var(--charcoal-color-container-hover-a);
}
.charcoal-icon-button[data-variant='Default'][data-active='false']:not(
    :disabled
  ):not([aria-disabled]):active,
.charcoal-icon-button[data-variant='Default'][data-active='false'][aria-disabled='false']:active {
  color: var(--charcoal-color-icon-tertiary-press);
  background-color: var(--charcoal-color-container-press-a);
}

.charcoal-icon-button[data-variant='Overlay'] {
  color: var(--charcoal-color-icon-on-on-img-default);
  background-color: var(--charcoal-color-container-on-img-default);
}
.charcoal-icon-button[data-variant='Overlay'][data-active='true']:not(
    :disabled
  ):not([aria-disabled]),
.charcoal-icon-button[data-variant='Overlay'][data-active='true'][aria-disabled='false'] {
  color: var(--charcoal-color-icon-on-on-img-press);
  background-color: var(--charcoal-color-container-on-img-press);
}
.charcoal-icon-button[data-variant='Overlay'][data-active='false']:not(
    :disabled
  ):not([aria-disabled]):hover,
.charcoal-icon-button[data-variant='Overlay'][data-active='false'][aria-disabled='false']:hover {
  color: var(--charcoal-color-icon-on-on-img-hover);
  background-color: var(--charcoal-color-container-on-img-hover);
}
.charcoal-icon-button[data-variant='Overlay'][data-active='false']:not(
    :disabled
  ):not([aria-disabled]):active,
.charcoal-icon-button[data-variant='Overlay'][data-active='false'][aria-disabled='false']:active {
  color: var(--charcoal-color-icon-on-on-img-press);
  background-color: var(--charcoal-color-container-on-img-press);
}

.charcoal-icon-button:not(:disabled):not([aria-disabled]):focus,
.charcoal-icon-button[aria-disabled='false']:focus {
  outline: none;
  box-shadow: var(--charcoal-color-custom-focus-effect);
}
.charcoal-icon-button:not(:disabled):not([aria-disabled]):focus-visible,
.charcoal-icon-button[aria-disabled='false']:focus-visible {
  box-shadow: var(--charcoal-color-custom-focus-effect);
}
.charcoal-icon-button:not(:disabled):not([aria-disabled]):focus:not(
    :focus-visible
  ),
.charcoal-icon-button[aria-disabled='false']:focus:not(:focus-visible) {
  box-shadow: none;
}
