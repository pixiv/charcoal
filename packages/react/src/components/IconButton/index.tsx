import React from 'react'
import styled from 'styled-components'
import { theme } from '../../styled'
import Clickable, { ClickableElement, ClickableProps } from '../Clickable'
import type { KnownIconType } from '@charcoal/icons'

type Variant = 'Default' | 'Overlay'
type Size = 'XS' | 'S' | 'M'

interface StyledProps {
  readonly variant?: Variant
  readonly size?: Size
  readonly icon: keyof KnownIconType
}

export type IconButtonProps = StyledProps & ClickableProps

const IconButton = React.forwardRef<ClickableElement, IconButtonProps>(
  function IconButtonInner(
    { variant = 'Default', size = 'M', icon, ...rest }: IconButtonProps,
    ref
  ) {
    validateIconSize(size, icon)
    return (
      <StyledIconButton {...rest} ref={ref} variant={variant} size={size}>
        <pixiv-icon name={icon} />
      </StyledIconButton>
    )
  }
)

export default IconButton

const StyledIconButton = styled(Clickable).attrs<
  Required<StyledProps>,
  ReturnType<typeof styledProps>
>(styledProps)`
  user-select: none;

  width: ${(p) => p.width}px;
  height: ${(p) => p.height}px;
  display: flex;
  align-items: center;
  justify-content: center;

  ${({ font, background }) =>
    theme((o) => [
      o.font[font],
      o.bg[background].hover.press,
      o.disabled,
      o.borderRadius('oval'),
      o.outline.default.focus,
    ])}
`

function styledProps(props: Required<StyledProps>) {
  return {
    ...props,
    ...variantToProps(props.variant),
    ...sizeToProps(props.size),
  }
}

function variantToProps(variant: Variant) {
  switch (variant) {
    case 'Default':
      return { font: 'text3', background: 'transparent' } as const
    case 'Overlay':
      return { font: 'text5', background: 'surface4' } as const
  }
}

function sizeToProps(size: Size) {
  switch (size) {
    case 'XS':
      return {
        width: 20,
        height: 20,
      }
    case 'S':
      return {
        width: 32,
        height: 32,
      }
    case 'M':
      return {
        width: 40,
        height: 40,
      }
  }
}

/**
 * validates matches of size and icon
 */
function validateIconSize(size: Size, icon: keyof KnownIconType) {
  let requiredIconSize: string
  switch (size) {
    case 'XS':
      requiredIconSize = '16'
      break
    case 'S':
    case 'M':
      requiredIconSize = '24'
      break
  }
  // アイコン名は サイズ/名前
  const result = /^\d*/u.exec(icon)
  if (result == null) {
    throw new Error('Invalid icon name')
  }
  const [iconSize] = result
  if (iconSize !== requiredIconSize) {
    console.warn(
      `IconButton with size "${size}" expect icon size "${requiredIconSize}, but got "${iconSize}"`
    )
  }
}
