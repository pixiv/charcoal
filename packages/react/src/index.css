:root {
  /* charcoal custom css variables */
  --charcoal-color-custom-focus-effect: 0 0 0 4px rgba(0, 150, 250, 0.32);
  --charcoal-color-custom-error-effect: 0 0 0 4px rgba(255, 43, 0, 0.32);
}
