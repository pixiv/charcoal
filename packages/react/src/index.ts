export {
  default as ComponentAbstraction,
  useComponentAbstraction,
  type LinkProps,
} from './core/ComponentAbstraction'
export { SSRProvider } from './core/SSRProvider'
export { default as Button, type ButtonProps } from './components/Button'
export {
  default as Clickable,
  type ClickableProps,
  type ClickableElement,
} from './components/Clickable'
export {
  default as IconButton,
  type IconButtonProps,
} from './components/IconButton'
export {
  default as Radio,
  type RadioProps,
  RadioGroup,
  type RadioGroupProps,
} from './components/Radio'
export {
  default as Select,
  type SelectProps,
  SelectGroup,
  type SelectGroupProps,
} from './components/Select'
export { default as Switch, type SwitchProps } from './components/Switch'
export {
  default as TextField,
  type TextFieldProps,
  type SingleLineTextFieldProps,
  type MultiLineTextFieldProps,
} from './components/TextField'
export { default as Icon, type IconProps } from './components/Icon'
export { default as Modal, type ModalProps } from './components/Modal'
export {
  default as LoadingSpinner,
  LoadingSpinnerIcon,
} from './components/LoadingSpinner'
export {
  default as SegmentedControl,
  type SegmentedControlProps,
} from './components/SegmentedControl'
export { default as Checkbox } from './components/Checkbox'
export { default as TagItem, type TagItemProps } from "./components/TagItem"
