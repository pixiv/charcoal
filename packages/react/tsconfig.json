{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "types": ["vitest/globals", "@testing-library/jest-dom"],
    "isolatedDeclarations": true,
  },
  "include": ["./src", "./docs", "./vitest.setup.ts"]
}
