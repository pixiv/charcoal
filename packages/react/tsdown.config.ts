import { defineConfig } from 'tsdown'
// import { styledComponentsPlugin } from '@charcoal-ui/esbuild-plugin-styled-components'
// import postcss from 'postcss'
// import fs from 'node:fs/promises'
// import path from 'node:path'

export default defineConfig({
  entry: ['src/index.ts'],
  format: ['esm', 'cjs'],
  // outExtension({ format }) {
  //   return {
  //     js: `.${format}.js`,
  //   }
  // },
  // banner: {
  //   js: `"use client";`,
  // },
  target: 'esnext',
  sourcemap: true,
  dts:true,
  // minify:true,
  
  // esbuildPlugins: [
  //   styledComponentsPlugin,
  //   {
  //     name: 'at-layer-charcoal',
  //     setup(build) {
  //       // NOTE: this will be called twice for esm and cjs.
  //       build.onEnd(async (result) => {
  //         const originalCssOutput = result.outputFiles?.find((file) =>
  //           file.path.endsWith('css')
  //         )
  //         if (!originalCssOutput) {
  //           throw new Error('[at-layer-charcoal]: expect original css output')
  //         }

  //         const layeredCssFilePath = originalCssOutput.path.replace(
  //           'index.css',
  //           'layered.css'
  //         )
  //         const layeredCssOutput = await postcss({
  //           postcssPlugin: 'at-layer-charcoal',
  //           Once(_, { AtRule, result }) {
  //             const atLayerNode = new AtRule({
  //               name: 'layer',
  //               params: 'charcoal',
  //               nodes: result.root.nodes,
  //             })
  //             result.root.nodes = [atLayerNode]
  //           },
  //         }).process(originalCssOutput.text, {
  //           from: originalCssOutput.path,
  //           to: layeredCssFilePath,
  //           map: {
  //             inline: false,
  //           },
  //         })

  //         // https://github.com/evanw/esbuild/issues/2999
  //         // this will be called before dist is created
  //         await fs.mkdir(path.dirname(layeredCssFilePath), { recursive: true })
  //         await Promise.all([
  //           fs.writeFile(layeredCssFilePath, layeredCssOutput.content),
  //           fs.writeFile(
  //             `${layeredCssFilePath}.map`,
  //             layeredCssOutput.map.toString()
  //           ),
  //         ])

  //         // eslint-disable-next-line no-console
  //         console.log(
  //           `${build.initialOptions.format?.toUpperCase()} dist/layered.css`
  //         )
  //       })
  //     },
  //   },
  // ],
})
