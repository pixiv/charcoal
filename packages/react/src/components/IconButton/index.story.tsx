import '@charcoal-ui/icons'
import IconButton from '.'
import { Meta, StoryObj } from '@storybook/react'

export default {
  title: 'IconButton',
  component: IconButton,
  parameters: {
    layout: 'centered',
  },
} as Meta<typeof IconButton>

export const Default: StoryObj<typeof IconButton> = {
  args: {
    icon: '16/Add',
  },
  render: (props) => {
    const title = props.icon.split('/').at(1)?.toLocaleLowerCase()

    return <IconButton title={title} {...props} />
  },
}

export const IsActive: StoryObj<typeof IconButton> = {
  render: () => {
    return <IconButton icon="16/Add" isActive />
  },
}

export const Overlay: StoryObj<typeof IconButton> = {
  render: () => {
    return <IconButton icon="16/Add" variant="Overlay" />
  },
}
