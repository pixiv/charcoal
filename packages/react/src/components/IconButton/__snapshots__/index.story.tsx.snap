// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots IconButton Default M 1`] = `
.c0 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 40px;
  height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: var(--charcoal-text3);
  background-color: var(--charcoal-transparent);
  border-radius: 999999px;
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
}

.c1:not(:disabled):not([aria-disabled]):hover,
.c1[aria-disabled='false']:hover {
  background-color: var( --charcoal-transparent-hover );
}

.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled='false']:active {
  background-color: var( --charcoal-transparent-press );
}

.c1:not(:disabled):not([aria-disabled]):focus,
.c1[aria-disabled='false']:focus,
.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled='false']:active,
.c1:not(:disabled):not([aria-disabled]):focus-visible,
.c1[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
}

<div
  data-dark={false}
>
  <button
    className="c0 c1"
    title="close"
  >
    <pixiv-icon
      name="24/Close"
    />
  </button>
</div>
`;

exports[`Storyshots IconButton Overlay M 1`] = `
.c0 {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled=false]) {
  cursor: default;
}

.c0:focus {
  outline: none;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 40px;
  height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: var(--charcoal-text5);
  background-color: var(--charcoal-surface4);
  border-radius: 999999px;
  -webkit-transition: 0.2s background-color,0.2s box-shadow;
  transition: 0.2s background-color,0.2s box-shadow;
}

.c1:not(:disabled):not([aria-disabled]):hover,
.c1[aria-disabled='false']:hover {
  background-color: var( --charcoal-surface4-hover );
}

.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled='false']:active {
  background-color: var( --charcoal-surface4-press );
}

.c1:not(:disabled):not([aria-disabled]):focus,
.c1[aria-disabled='false']:focus,
.c1:not(:disabled):not([aria-disabled]):active,
.c1[aria-disabled='false']:active,
.c1:not(:disabled):not([aria-disabled]):focus-visible,
.c1[aria-disabled='false']:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1:disabled,
.c1[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
}

<div
  data-dark={false}
>
  <button
    className="c0 c1"
    title="close"
  >
    <pixiv-icon
      name="24/Close"
    />
  </button>
</div>
`;
